### DISCUSSION TITLE
Custom left and right margin marker

### DISCUSSION THREAD

**Discussion by CDucloux** [typst/typst#4781] [Q&A]
In my typst document, I'm currently trying to implement a system of rectangles in the margin with the text of the current section in it.

As for now, I have the following, which is not too far from what I want in the end: 

<img width="563" alt="custom_marker" src="https://github.com/user-attachments/assets/01463cc5-fb2d-4169-b1ee-0f4a88da309b">

The code I use is the following:

```typ
#let indicator = place(
  dx:38em,
  dy:5em,
  rotate(
    90deg,
    rect(
      [#align(center,v(1.25em) + text(context {
let elems = query(
    selector(heading).before(here())
  )
  upper(elems.last().body)
}
, fill:white))],
      fill: black,
      width: 140pt,
      height: 40pt,
      radius: (
    bottom-left: 10pt,
    bottom-right: 10pt,
  ))))
```

There are some problems, though:

- The `place` function is relative, meaning that depending on where I put the `#indicator` function on the page, it won't be displayed at the same height - **is there a way to put the marker on all pages rather than doing it manually ?**
- I would like to be able to put the marker on left/right margin depending on odd/even page
- The color should also change when the name of the marker changes

Thank you for your time!

**Comment by Mc-Zen** [comment]
Hi @CDucloux ,
you could actually place it in the header or footer!

I'm just posting a rough version of how that works - you will want to tweak it to your needs. 
```typ
#let get-current-heading() = {
  let loc = here()
  let heading-selector = heading.where(level: 1, outlined: true) // Customize!
  let el = query(heading-selector.after(loc)).at(0, default: none)
  if el != none and el.location().page() == loc.page() {
    return el
  } else {
    return query(heading-selector.before(loc)).at(-1, default: none)
  }
}
  
#set page(header: context {
  let latest-heading = get-current-heading()
  if latest-heading == none { return }
  let content = text(1.4em, latest-heading)
  place(
    right + horizon, 
    dx: 3.4cm, // depends on page's right margin
    dy: page.height / 2, 
    rotate(90deg, rect(fill: yellow, inset: 1em, content))
  )
})
```

Let me shortly lay down what happens here. We set the `page.header` and require context so we can query for the latest heading. Doing this from the header is not entirely trivial (because a new heading might start on this page which is then always _after_ the page header), so we have a helper function for that. If there is no heading we skip (might be the case on the title page, table-of-contents etc.). 

Then we place some content `right + horizon`-aligned on the page, shifting the vertical position by half of the page. Note that this will not put the content exactly at the page middle because of the margins. The field `page.header-ascent` also plays into that. If you require perfect positioning for all page sizes, margins, and header-ascent values, you need to put in a little more effort but it is definitely possible - just not worth the effort maybe. 

For the same reason I chose to do the `dx` shift manually. You can access `page.margin` but that can have different formats or be just `auto` - all in all a lot of cases to handle :D  

Let me know if that helps!
